### PR TITLE
Initial version of protoc-gen-jsonschema

### PR DESCRIPTION
**Pull request by morphar**
Hi @timburks,

Here is the initial version of the JSON Schema generator.
Let me know what you think :)

I have tried to keep it simple and consistent with gnostic, but I may have missed something of course.

Currently every proto message gets it's own file, but I don't think it will be a problem to enable single file generation with every schema described in "definitions".

### CODE REVIEW

**Review by timburks — APPROVED** [review]
Thanks, this looks good to me. Could you elaborate on the uses for this? For example, is there software that automatically converts JSON (following these schemas) to protos in a similar way to how [HTTP and gRPC Transcoding](https://google.aip.dev/127) works?

I'm also a little concerned that this might be behind the latest [JSON Schema specification](https://json-schema.org/), but that's a more general problem affecting all of the JSON schema-handling in gnostic (so it's not a blocker here).

File: apps/protoc-gen-jsonschema/generator/json-schema.go [hunk 1]
```diff
@@ -0,0 +1,352 @@
+// Copyright 2020 Google LLC. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (morphar) [reply apps/protoc-gen-jsonschema/generator/json-schema.go:1]: 👍 I will fix this

**Comment by morphar** [comment]
Personally I have 2 use cases for this:
1. Expose some of our internal structure as JSON Schemas, for cases where proto is not used by clients
2. Query generators based on a couple of different "languages", where I need to know the type of values. This is done by relating values to JSON Schemas as this is much faster than reflecting in Go.

The last use case also deals with user generated JSON Schemas, which can reference our pre-defined JSON Schemas.

As we talked about earlier, I'm not entirely sure this is a perfect fit for gnostic, as this is more of a one-way thing as services is not a part of the result.
It makes a lot of sense in my context, but that may be too special a case?

File: apps/protoc-gen-jsonschema/generator/json-schema.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (morphar) [reply apps/protoc-gen-jsonschema/generator/json-schema.go]: Done

**Comment by morphar** [comment]
BTW... With regards to the schema version, I have made it possible to set the version from options and made changes to the jsonschema code, so that it differentiates between id (draft-04) and $id (draft-06+), based on the version.
As far as I can remember, there should not be anything else, that is not compatible all the way from draft-04 and up to the latest 2020-12.

**Comment by morphar** [comment]
@timburks Just a heads-up: As you may have seen, I have created another pull request (#276), which doesn't contain this code, so if this or #276 is merged, the other probably needs to merge master again, before it can be safely merged.

**Comment by morphar** [comment]
@timburks what are your thoughts on this?
Should it be merged or should I start it up as a separate project?

**Comment by timburks** [comment]
Thanks for your updates and discussion. I'm happy to merge this, but an important thing to keep in mind is that contributing this to gnostic has ownership implications - if you put this in a separate repo you have more control, including the ability to assert copyright and potentially relicense. gnostic is Apache-licensed, so I don't think there would be anything blocking you from moving it out of this repo (if it was merged here) and into a separate repo in the future, but you might need to retain the Google copyright notices.

If that's still ok with you, reply to confirm and I'll merge this.

**Comment by morphar** [comment]
I'm totally ok with that.
The code is already a copy of code from gnostic itself and I'd rather contribute to OS and have a coherent library, than have ownership over something that has less eyes on it :)

**Comment by timburks** [comment]
@morphar Thank you!

**Comment by morphar** [comment]
And thank you @timburks! :)

**Pull request merged**